## Q1] 
An auto manufacturer uses 500 tons of steel per day. Each order incurs a fixed cost of \\$2,250.
Suppose that the steel supplier offers the auto manufacturer a price of $1,490 per ton of steel
if Q < 1200 tons; \\$1,220 per ton of steel if 1200 ≤ Q < 2400 tons; and \\$1,100 per ton of steel
if Q ≥ 2400 tons. The annual holding cost rate, i is 0.25

(a) Calculate Q∗ for the all units discount structure

(b) Calculate the average total cost that the company will incur with the optimal order quantity you calculated above

In [11]:
# Given data 

import numpy as np

# Daily demand D = 500 tons
# Annual Demand D * 365
D = 500 * 365

# Fixed cost (ordering cost) 
S = 2250

# Holding cost rate 
i = 0.25

# Holding cost H = i * P

#1  if Q < 1200 tons, supplier offers = $ 1490 per ton of steel
P1 = 1490
H1 = i * P1
print('The holding cost is',H1)
def EOQ(S, D, H1):

    if(S>0 and D>0 and H1>0):

        Q1 = round((np.sqrt(2*S*D/H1)))
        AOC1 = D*S/Q1
        AHC1 = Q1*H1/2
        PUC1 = P1 * D
        ATC1 = AOC1+AHC1+PUC1 # PUC is the purchase cost
        print('The economic order quantity is =', Q1)
        print('The average annual cost is',ATC1)
        if Q1 > 1119:
            print('This quantity is out of given range')
        else:
            print('This quantity is within given range')     
EOQ(S,D,H1)
print()
#2 if 1200 =< Q < 2400, supplier offers = $ 1220 per ton of steel
P2 = 1220
H2 = i * P2
print('The holding cost is',H2)
def EOQ(S, D, H2):

    if(S>0 and D>0 and H2>0):

        Q2 = round((np.sqrt(2*S*D/H2)))
        AOC2 = D*S/Q2
        AHC2 = Q2*H2/2
        PUC2 = P2 * D 
        ATC2 = AOC2+AHC2+PUC2 # PUC is the purchase cost
        print('The economic order quantity is =', Q2)
        print('The average annual cost is',ATC2)
        if Q2 > 2400:
            print('This quantity is out of the given range')
        elif Q2 < 1200:
            print('This quantity is out of the given range')
        else:
            print('This quantity is within the given range')
EOQ(S,D,H2)
print()
#3 if Q >= 2400, supplier offers = $ 1100 per ton of steel
P3 = 1100
H3 = i * P3
print('The holding cost is',H3)
def EOQ(S, D, H3):

    if(S>0 and D>0 and H3>0):

        Q3 = round((np.sqrt(2*S*D/H3)))
        AOC3 = D*S/Q3
        AHC3 = Q3*H3/2
        PUC3 = P3 * D
        ATC3 = AOC3+AHC3+PUC3 # PUC is the purchase cost
        print('The economic order quantity is =', Q3)
        print('The average annual cost is',ATC3)
        if Q3 >= 2400:
            print('This quantity is within the given range')
        else:
            print('This quantity is out of the given range')
EOQ(S,D,H3)  

The holding cost is 372.5
The economic order quantity is = 1485
The average annual cost is 272478096.4015151
This quantity is out of given range

The holding cost is 305.0
The economic order quantity is = 1641
The average annual cost is 223150481.01919562
This quantity is within the given range

The holding cost is 275.0
The economic order quantity is = 1728
The average annual cost is 201225230.20833334
This quantity is out of the given range


Comparing the 3 values of total annual cost, we can see that the minimum value is obtained for a quantity of Q >= 2400 so optimal quantity is 2400 tons. 

# Q2]
We will answer the same question under the incremental discount structure. An auto manufacturer uses 500 tons of steel per day. Each order incurs a fixed cost of \\$2,250. Suppose that the steel supplier offers the auto manufacturer a price of \\$1,490 per ton of steel if Q < 1200
tons; \\$1,220 per ton of steel if 1200 ≤ Q < 2400 tons; and \\$1,100 per ton of steel if Q ≥ 2400
tons. The annual holding cost rate, i is 0.25.

(a) Calculate Q∗ for the incremental discount structure

(b) Calculate the optimal average total cost per year for the incremental discount structure.

In [6]:
# Given data 

import numpy as np

# Daily demand D = 500 tons
# Annual Demand D * 365
D = 500 * 365

# Fixed cost (ordering cost) 
S = 2250

# Holding cost rate 
i = 0.25

# Holding cost H = i * P

#if Q<1200 tons
c1 = 1490
#if 1200<=Q<2400 tons
c2 = 1220
 #if Q >= 2400 tons
c3 = 1100 

#For incremental discount ordering cost changes
S1 = S                    #if Q<1200 tons
S2 = S + 324000           #if 1200<=Q<2400 tons
S3 = S + 612000           #if Q >= 2400 tons

def EOQ(K,D,i,c):
    Q=np.sqrt(2*S*D/(i*c))
    return Q

Q1 = EOQ(S1,D,i,c1)
print('Optimal order quantity for incremental discount if Q<1200 is', Q1)

Q2 = EOQ(S2,D,i,c2)
print('Optimal order quantity for incremental discount if 1200<=Q<2400 is', Q2)

Q3 = EOQ(S3,D,i,c3)
print('Optimal order quantity for incremental discount if Q >= 2400 is', Q3)

#As Q1 should be <1200 tons
Q1 = 1200        
Q3 = 28553

C1_new = 1490
C3_new = 1100 + 612000/Q3

# Average total cost
def TC(S,D,i,c,Q):
    C=(i*c*Q*0.5)+(S*D)/Q + (c*D)
    return C

AvgCostQ1 = TC(S,D,i,C1_new,Q1)
print('Average total cost for incremental discount at optimal order quantity',Q1,'tons is $',AvgCostQ1)

AvgCostQ3 = TC(S,D,i,C3_new,Q3)
print('Average total cost for incremental discount at optimal order quantity',Q3,'tons is $',AvgCostQ3)


Optimal order quantity for incremental discount if Q<1200 is 1484.822543800161
Optimal order quantity for incremental discount if 1200<=Q<2400 is 1640.9213725281513
Optimal order quantity for incremental discount if Q >= 2400 is 1728.1098449935514
Average total cost for incremental discount at optimal order quantity 1200 tons is $ 272490687.5
Average total cost for incremental discount at optimal order quantity 28553 tons is $ 208678591.67994606


From above analysis, the best optimal quantity is 28553 tons with total cost of \\$ 208678591.68 because it is the lowest.

# Q3]
An auto manufacturer uses 500 tons of steel per day. The company pays \\$1,100 per ton of steel purchased, and each order incurs a fixed cost of \\$2,250. The annual holding cost rate, i is 0.25.
Now we will try to see if there are any benefits to allowing shortages for this company. Note that this company uses steel for a manufacturing process. The shortages will be treated as backorders since the auto manufacturer will wait for the steel to be delivered. As such, shortages of this
supply results in the production line to halt so there may be big consequence to shortages. (In the case of this manufacturer, luckily, the production line is used to manufacture another model so during the backorder period the line is used for something else.) The management is estimating that the backorder cost, b per unit per unit time is \\$475.

(a) Calculate the optimal ordering quantity, Q∗

(b) Calculate the optimal backorder quantity, B∗

(c) What is the overall cost of this optimal policy you have calculated?

(d) How much lower is the above calculated cost in comparison to the total annual cost you would have obtained if you did not allow shortages (i.e., backorders)?

In [3]:
# Given data 

import numpy as np

# Daily demand D = 500 tons
# Annual Demand D * 365
D = 500 * 365

# Fixed cost (ordering cost) 
S = 2250

# Holding cost rate 
i = 0.25

c = 1100

H = i * c

# Backorder cost
b = 475

def EOQ(S, D, H):

    if(S>0 and D>0 and H>0):

        Q = round((np.sqrt((2*S*D)/H))*(np.sqrt((H+b)/b)))
        B = round((np.sqrt((2*S*D)/b))*(np.sqrt(H/(H+b))))
        C1 = (D*S)/Q
        C2 = c * D 
        C3 = (((Q-B)**2)*H)/(2*Q)
        C4 = ((B**2)*b)/(2*Q)
        C5 = (H*Q)/2
        OC1 = round(C1 + C2 + C3 + C4)
        OC2 = round(C1 + C2 + C5)
        diff = OC2 - OC1
        print(' (a) The optimal order quantity is =', Q)
        print(' (b) The optimal backorder quantity is =', B)
        print(' (c) The overall cost of this optimal policy is',OC1)
        print('The overall cost of this policy without backorder is',OC2)
        print(' (d) The above calculated cost would be',diff,'lower than the cost if shortages are not allowed')
EOQ(S,D,H)

 (a) The optimal order quantity is = 2171
 (b) The optimal backorder quantity is = 796
 (c) The overall cost of this optimal policy is 201128199
The overall cost of this policy without backorder is 201237653
 (d) The above calculated cost would be 109454 lower than the cost if shortages are not allowed


# Q4]
Consider the same question again. Suppose that instead of an auto manufacturer, you are managing the inventory of a retailer, where the implication of a shortage is a unit of lost sale. Assume all other parameters are the same: c = \\$1100, D = 500 tons per day, K = \\$2250 per order, i = 0.25 \\$/unit/year. For what value of the unit shortage penalty, p, would you decide to satisfy all demand using the EOQ policy?


In [4]:
# Given data 

import numpy as np

# Daily demand D = 500 tons
# Annual Demand D * 365
D = 500 * 365

# Fixed cost (ordering cost) 
S = 2250

# Holding cost rate 
i = 0.25
c = 1100
H = i * c

# It is also given that we have to satisfy all of the demand, therefore x = 0
# Therefore we will be using the condition D*(p-c) > np.sqrt(2*S*D*H), where the slope is positive C is increasing so it is optimal to set x*=0

def EOQ(S, D, H):

    if(S>0 and D>0 and H>0):

        Q = round((np.sqrt(2*S*D/H)))
        print('The economic order quantity is',Q)
EOQ(S,D,H)    
p = round(((np.sqrt(2*S*D*H))/D)+c)
print('The value of the unit shortage penalty since the condition states that we need to satisfy all demand p >',p)

The economic order quantity is 1728
The value of the unit shortage penalty since the condition states that we need to satisfy all demand p > 1103
